src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,66 +25,164 @@
+    public void setState(String projectKey, String slug, String serverId, String fromBranch, String toBranch, BitbucketPullState state) {
+        PullRequestStoreImpl.CacheKey cacheKey =
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        pullRequests.getOrDefault(cacheKey, new RepositoryStore()).setState(fromBranch, toBranch, state);
     }
 
     @Override
     public boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository repository) {
-
         PullRequestStoreImpl.CacheKey key =
                 new PullRequestStoreImpl.CacheKey(repository.getProjectKey(), repository.getRepositorySlug(), repository.getServerId());
-        return pullRequests.getOrDefault(key, new ConcurrentLinkedQueue<>())
-                .stream()
-                .filter(pullRequest -> pullRequest.getFromRef().getDisplayId().equals(branchName))
-                .findFirst()
-                .isPresent();
+        return pullRequests.getOrDefault(key, new RepositoryStore()).hasOpenPullRequests(branchName);
     }
 
     @Override
-    public Optional<BitbucketPullRequest> getPullRequest(String key, String slug, String serverId, int pullRequestId) {
+    public Optional<MinimalPullRequest> getPullRequest(String projectKey, String slug, String serverId, String fromBranch, String toBranch) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest == null || pullRequest.isEmpty()) {
-            return Optional.empty();
-        }
-        return pullRequest.stream().filter(pr -> pr.getId() == pullRequestId).findFirst();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return Optional.ofNullable(pullRequests.getOrDefault(cacheKey, new RepositoryStore())
+                .getPullRequest(fromBranch, toBranch));
     }
 
     @Override
-    public void refreshStore(String key, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
+    public void refreshStore(String projectKey, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest != null) {
-            pullRequest.clear();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        long oldestUpdateInStore = pullRequests.getOrDefault(cacheKey, new RepositoryStore()).findOldestUpdateInStore();
+
+        // update/add pull requests only with information that is newer than the oldest update
+        bbsPullRequests.filter(pr -> pr.getUpdatedDate() > oldestUpdateInStore).forEach(pr -> updatePullRequest(serverId, pr));
```

Comment:
this doesn't work for us. This version of the code would iterate over _every single PR_ in the remote BbS. We can't do that. We must stop the stream once we reach the date. The old code did this

```java
 //bbspullrequests is ordered from newest to oldest
//noinspection ResultOfMethodCallIgnored
        bbsPullRequests.map(pr -> {
         //check if the PR needs to be updated, and if so update.
            if (pr.getUpdatedDate() > oldestUpdate.get()) {
                updatePullRequest(serverId, pr);
            }
//return the date for the next part of our stream
            return pr.getUpdatedDate();     })
         //only return dates that are older than our oldest so find first works
.filter(updated -> updated < oldestUpdate.get())
 //findFirst will find the first result of the stream and terminate the stream, thus once we've found an older PR we will stop streaming.
                .findFirst();

```
it is a bit wonky, as we do not use the result of the `findFirst`. I added `//noinspection ResultOfMethodCallIgnored` to help indicate that it is intentional and not have tooling being 'helpful' to suggest ways to do it differently.